pkg/oidc/oidc.go

```diff
@@ -0,0 +1,351 @@
+		return TokenProcessor{}, fmt.Errorf("%s issuer is not trusted", issuer)
+	}
+	tokenProcessor.issuer = issuer
+	logger.Debugw("Added trusted issuer to TokenProcessor", "issuer", tokenProcessor.issuer)
+
+	logger.Debugw("Applying TokenProcessorOptions")
+	for _, option := range options {
+		err := option(&tokenProcessor)
+		if err != nil {
+			return TokenProcessor{}, fmt.Errorf("failed to apply TokenProcessorOption: %w", err)
+		}
+	}
+	logger.Debugw("Applied all TokenProcessorOptions")
+
+	logger.Debugw("Created token processor", "issuer", tokenProcessor.issuer)
+	return tokenProcessor, nil
+}
+
+// tokenIssuer gets the issuer from the token.
+// It doesn't verify the token, just parses its claims.
+// It's used to create a new TokenProcessor.
+// TODO: signAlgorithm should be a TokenProcessor field.
+// TODO: should we abstract usage of jwt and go-jose libraries?
+func (tokenProcessor *TokenProcessor) tokenIssuer(signAlgorithm []string) (string, error) {
+	logger := tokenProcessor.logger
+	logger.Debugw("Getting issuer from token")
+	// claims := jwt.Claims{}
+	claims := Claims{}
+	var signAlgs []jose.SignatureAlgorithm
+	for _, alg := range signAlgorithm {
+		signAlgs = append(signAlgs, jose.SignatureAlgorithm(alg))
+		logger.Debugw("Added sign algorithm to token processor", "signAlgorithm", alg)
+	}
+	logger.Debugw("Added sign algorithms to token processor")
+	// TODO(dekiel) research if we can use jwt.DecodeSegment instead of jwt.ParseSigned
+	//  to avoid parsing the token twice.
+	parsedJWT, err := jwt.ParseSigned(tokenProcessor.rawToken, signAlgs)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse oidc token: %w", err)
+	}
+	logger.Debugw("Parsed oidc token")
+	err = parsedJWT.UnsafeClaimsWithoutVerification(&claims)
+	if err != nil {
+		return "", fmt.Errorf("failed to get claims from unverified token: %w", err)
+	}
+	logger.Debugw("Got claims from unverified token", "claims", fmt.Sprintf("%+v", claims))
+	return claims.Issuer, nil
+}
+
+// isTrustedIssuer checks if the issuer is trusted.
+// It's used to create a new TokenProcessor.
+func (tokenProcessor *TokenProcessor) isTrustedIssuer(issuer string, trustedIssuers map[string]Issuer) bool {
+	logger := tokenProcessor.logger
+	logger.Debugw("Checking if issuer is trusted", "issuer", issuer)
+	logger.Debugw("Trusted issuers", "trustedIssuers", trustedIssuers)
+	if _, exists := trustedIssuers[issuer]; exists {
+		logger.Debugw("Issuer is trusted", "issuer", issuer)
+		return true
+	}
+	logger.Debugw("Issuer is not trusted", "issuer", issuer)
+	return false
+}
+
+// Issuer returns the issuer of the token.
+func (tokenProcessor *TokenProcessor) Issuer() string {
+	return tokenProcessor.issuer
+}
+
+// Claims verify and parse the token to get the token claims.
+// It unmarshal the claims into the provided claims struct.
+func (tokenProcessor *TokenProcessor) Claims(ctx context.Context, verifier TokenVerifierInterface, claims ClaimsInterface) error {
```

Comment:
Change the name of this function to sth like, "VerifyAndExtractClaims".

Reply:
renamed